File: backend/src/main/java/com/sesjob/entity/Job.java
package com.sesjob.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

/**
 * 案件エンティティ
 * SES/フリーランス案件の情報を保持
 */
@Entity
@Table(name = "jobs", uniqueConstraints = @UniqueConstraint(columnNames = { "source", "source_url" }), indexes = {
        @Index(name = "idx_jobs_status", columnList = "status"),
        @Index(name = "idx_jobs_source", columnList = "source"),
        @Index(name = "idx_jobs_posted_at", columnList = "posted_at"),
        @Index(name = "idx_jobs_max_price", columnList = "max_price")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Job {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // ソース情報
    @Column(name = "source", nullable = false, length = 50)
    private String source;

    @Column(name = "source_url", nullable = false, length = 500)
    private String sourceUrl;

    @Column(name = "source_id", length = 100)
    private String sourceId;

    @Column(name = "title", nullable = false, length = 300)
    private String title;

    // 報酬情報
    @Column(name = "min_price")
    private Integer minPrice;

    @Column(name = "max_price")
    private Integer maxPrice;

    @Enumerated(EnumType.STRING)
    @Column(name = "price_type", length = 20)
    private PriceType priceType;

    @Column(name = "settlement_hours", length = 50)
    private String settlementHours;

    // スキル要件
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "required_skills", columnDefinition = "json")
    private List<String> requiredSkills;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "preferred_skills", columnDefinition = "json")
    private List<String> preferredSkills;

    @Column(name = "experience_years", length = 50)
    private String experienceYears;

    // 勤務条件
    @Column(name = "location", length = 100)
    private String location;

    @Enumerated(EnumType.STRING)
    @Column(name = "remote_type", length = 20)
    private RemoteType remoteType;

    @Column(name = "work_days", length = 50)
    private String workDays;

    @Column(name = "start_date")
    private LocalDate startDate;

    @Column(name = "contract_period", length = 100)
    private String contractPeriod;

    // その他
    @Column(name = "company_name", length = 200)
    private String companyName;

    @Column(name = "industry", length = 100)
    private String industry;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    // メタ情報
    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20)
    @Builder.Default
    private JobStatus status = JobStatus.NEW;

    @Column(name = "is_favorite")
    @Builder.Default
    private Boolean isFavorite = false;

    @Column(name = "posted_at")
    private LocalDateTime postedAt;

    @Column(name = "crawled_at", nullable = false)
    private LocalDateTime crawledAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "created_at")
    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();

    @PreUpdate
    public void preUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    // Enums
    public enum PriceType {
        MONTHLY, HOURLY
    }

    public enum RemoteType {
        FULL, PARTIAL, NONE
    }

    public enum JobStatus {
        NEW, READ, APPLIED, CLOSED
    }
}

File: crawler/models/job.py
"""
SES Job Crawler - Data Models
"""
from pydantic import BaseModel, Field, HttpUrl
from typing import Optional, List
from datetime import date, datetime
from enum import Enum


class PriceType(str, Enum):
    MONTHLY = "MONTHLY"
    HOURLY = "HOURLY"


class RemoteType(str, Enum):
    FULL = "FULL"
    PARTIAL = "PARTIAL"
    NONE = "NONE"


class JobData(BaseModel):
    """案件データモデル"""
    
    source: str = Field(..., description="データソース名")
    source_url: HttpUrl = Field(..., description="元のURL")
    source_id: Optional[str] = Field(None, description="サイト内のID")
    title: str = Field(..., description="案件タイトル")
    
    # 報酬情報
    min_price: Optional[int] = Field(None, description="最低単価（万円）")
    max_price: Optional[int] = Field(None, description="最高単価（万円）")
    price_type: Optional[PriceType] = Field(None, description="単価種別")
    settlement_hours: Optional[str] = Field(None, description="精算幅")
    
    # スキル要件
    required_skills: List[str] = Field(default_factory=list, description="必須スキル")
    preferred_skills: List[str] = Field(default_factory=list, description="歓迎スキル")
    experience_years: Optional[str] = Field(None, description="経験年数")
    
    # 勤務条件
    location: Optional[str] = Field(None, description="勤務地")
    remote_type: Optional[RemoteType] = Field(None, description="リモート可否")
    work_days: Optional[str] = Field(None, description="稼働日数")
    start_date: Optional[date] = Field(None, description="稼働開始日")
    contract_period: Optional[str] = Field(None, description="契約期間")
    
    # その他
    company_name: Optional[str] = Field(None, description="企業名")
    industry: Optional[str] = Field(None, description="業界")
    description: Optional[str] = Field(None, description="詳細説明")
    posted_at: Optional[datetime] = Field(None, description="掲載日時")

    def to_api_request(self) -> dict:
        """API送信用の辞書に変換"""
        data = self.model_dump(exclude_none=True)
        # Convert HttpUrl to string
        data["source_url"] = str(data["source_url"])
        # Convert snake_case to camelCase for API
        return self._to_camel_case(data)
    
    def _to_camel_case(self, data: dict) -> dict:
        """snake_case を camelCase に変換"""
        def convert_key(key: str) -> str:
            components = key.split('_')
            return components[0] + ''.join(x.title() for x in components[1:])
        
        result = {}
        for key, value in data.items():
            new_key = convert_key(key)
            if isinstance(value, dict):
                result[new_key] = self._to_camel_case(value)
            elif isinstance(value, list):
                result[new_key] = value
            elif isinstance(value, Enum):
                result[new_key] = value.value
            elif isinstance(value, (date, datetime)):
                result[new_key] = value.isoformat() if value else None
            else:
                result[new_key] = value
        return result


class CrawlResult(BaseModel):
    """クロール結果"""
    source: str
    success: bool
    jobs_count: int
    error_message: Optional[str] = None
    crawled_at: datetime = Field(default_factory=datetime.now)
